components/notebook-controller/controllers/notebook_controller.go

```diff
@@ -449,6 +473,39 @@ func generateVirtualService(instance *v1beta1.Notebook) (*unstructured.Unstructu
 			"timeout": "300s",
 		},
 	}
+
+	// Create a template map for the headers section
+	// of the Istio VirtualService
+	headers := map[string]interface{}{
+		"request": map[string]interface{}{
+			"set": map[string]interface{}{},
+		},
+	}
+
+	// Blacklist for request headers that are not allowed to be set
+	blacklist := []string{"userid-header", "X-Forwarded-Proto"}
+
+	// If key "notebooks.kubeflow.org/http-headers-request-set" is present on the notebook resource,
+	// split the literal string into keys and values and add them
+	// to the headers variable. Finally, add the headers section to
+	// the variable with the http spec.
+	if _, ok := annotations["notebooks.kubeflow.org/http-headers-request-set"]; ok && len(annotations["notebooks.kubeflow.org/http-headers-request-set"]) > 0 {
+		requestHeaders := strings.Split(annotations["notebooks.kubeflow.org/http-headers-request-set"], "\n")
+		for _, kv := range requestHeaders {
+			if len(strings.Split(kv, ": ")) == 2 {
+				k := strings.Split(kv, ": ")[0]
+				v := strings.Split(kv, ": ")[1]
+				_, key := Find(blacklist, k)
+				_, value := Find(blacklist, v)
+				if !key && !value {
```

Comment:
I'm not familiar with this syntax in Go and haven't seen in elsewhere. Can you rewrite this statement to make it more clear what it is you're checking?

Reply:
Not sure how to rewrite it. It is using the `Find` function and only continues if the value `k` and `v` are not in the blacklist variable. I'm going to rewrite this to be a whitelist instead, but more that that I am not sure how I would do it.

Reply:
I got confused because of the variable names. I think you should rewrite it like this:
```go
if exists(blacklist, k) {
  continue
}
```
to essentially rule out the bad case and then continue.

BTW, I saw you check both the key and value. Is this right? Shouldn't we check on the header name?